crates/rpc-types-engine/src/payload.rs

```diff
@@ -0,0 +1,175 @@
+
+        let signature = Signature::try_from(sig_data)?;
+        let hash = PayloadHash::from(block_data);
+
+        let payload = ExecutionPayload::from(ExecutionPayloadV1::from_ssz_bytes(block_data)?);
+
+        Ok(Self { payload, signature, payload_hash: hash, parent_beacon_block_root: None })
+    }
+
+    /// Decode a payload envelope from a snappy-compressed byte array.
+    /// The payload version decoded is `ExecutionPayloadV2` from SSZ bytes.
+    pub fn decode_v2(data: &[u8]) -> Result<Self, PayloadEnvelopeError> {
+        let mut decoder = snap::raw::Decoder::new();
+        let decompressed = decoder.decompress_vec(data)?;
+
+        if decompressed.len() < 66 {
+            return Err(PayloadEnvelopeError::InvalidLength);
+        }
+
+        let sig_data = &decompressed[..65];
+        let block_data = &decompressed[65..];
+
+        let signature = Signature::try_from(sig_data)?;
+        let hash = PayloadHash::from(block_data);
+
+        let payload = ExecutionPayload::from(ExecutionPayloadV2::from_ssz_bytes(block_data)?);
+
+        Ok(Self { payload, signature, payload_hash: hash, parent_beacon_block_root: None })
+    }
+
+    /// Decode a payload envelope from a snappy-compressed byte array.
+    /// The payload version decoded is `ExecutionPayloadV3` from SSZ bytes.
+    pub fn decode_v3(data: &[u8]) -> Result<Self, PayloadEnvelopeError> {
+        let mut decoder = snap::raw::Decoder::new();
+        let decompressed = decoder.decompress_vec(data)?;
+
+        if decompressed.len() < 98 {
+            return Err(PayloadEnvelopeError::InvalidLength);
+        }
+
+        let sig_data = &decompressed[..65];
+        let parent_beacon_block_root = &decompressed[65..97];
+        let block_data = &decompressed[97..];
+
+        let signature = Signature::try_from(sig_data)?;
+        let parent_beacon_block_root = Some(B256::from_slice(parent_beacon_block_root));
+        let hash = PayloadHash::from(block_data);
+
+        let payload = ExecutionPayload::from(ExecutionPayloadV3::from_ssz_bytes(block_data)?);
+
+        Ok(Self { payload, signature, payload_hash: hash, parent_beacon_block_root })
+    }
+}
+
+/// Represents the Keccak256 hash of the block
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+#[cfg_attr(any(test, feature = "arbitrary"), derive(arbitrary::Arbitrary))]
+pub struct PayloadHash(B256);
+
+impl From<&[u8]> for PayloadHash {
+    /// Returns the Keccak256 hash of a sequence of bytes
+    fn from(value: &[u8]) -> Self {
+        Self(keccak256(value))
+    }
+}
+
+impl PayloadHash {
+    /// The expected message that should be signed by the unsafe block signer.
+    pub fn signature_message(&self, chain_id: u64) -> B256 {
+        let domain = B256::ZERO.as_slice();
+        let chain_id = B256::left_padding_from(&chain_id.to_be_bytes()[..]);
+        let payload_hash = self.0.as_slice();
+        keccak256([domain, chain_id.as_slice(), payload_hash].concat())
+    }
+}
```

Comment:
I really thought we had this somewhere but I've looked and can't seem to find it. I'm ok adding it and refactoring it out later if we find it down the road.

Reply:
Ah, I took it from here: https://github.com/paradigmxyz/op-rs/blob/bdd68f73cce39643c285292be18a56f85f9a4c46/crates/net/src/types/payload.rs#L10